Ephesus runtime upgrade ci

dobertRowneySr opened the pull request:
Refers to #4555

This needs #4599 in order to have `joystream-node.compact.compressed.wasm` as runtime code for the proposal in order to avoid triggering a block weight exceeded error.

This also needs #4608 as it is based on top of that

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1202078489349362/1203822321386864) by [Unito](https://www.unito.io)

Commit:
chore: enable runtime-upgrade ci-checks

Commit:
feat: add dispatch when data migration is needed

Commit:
fix: remove root key value configuration for Sudo

Commit:
chore: remove unused NPoSSwitch test

Commit:
fix: migration workflows build joysteram/js

Commit:
fix: remove carthage scenario from CI

Commit:
fix: starting to fix runtime upgrade

Commit:
fix: runtime scripts

Commit:
fix modify docker file to start env

Commit:
Merge branch 'ephesus' into ephesus-runtime-upgrade-ci

Commit:
fix: runtime update script

Commit:
fix: re-enable sudo

Commit:
fix: runtime endpoints & edit forkof

Commit:
fix: add line for copying runtime wasm

Commit:
fix: set proper docker images

Commit:
fix: modify minisecret

Commit:
fix: pallet selection

Commit:
fix: revert uri path

Commit:
fix: tags

Commit:
fix: setting up the proper new wasm path

Commit:
fix: update ignored pallets

Commit:
fix: paths & formatting

Commit:
fix: linter

Commit:
fix: runtime upgrade

Commit:
fix: remove debug assertions

Commit:
fix: CI scripts

Commit:
fix: runtime profile env var

Commit:
Merge branch 'ephesus-merge-mainnet-lib' into ephesus-runtime-upgrade-ci

Commit:
fix: misc

Commit:
Merge branch 'ephesus' into ephesus-runtime-upgrade-ci

Commit:
fix: runtime compressed path

Commit:
Merge branch 'ephesus-CICD-fix' into ephesus-runtime-upgrade-ci

Commit:
fix: add main

Commit:
fix: set sudo config for master

Commit:
refactor: docker scripts

Commit:
fix: rename tags

Commit:
fix: remove runtime upgrade job from full

Commit:
fix: gh workflow

Commit:
fix: gh workflow

Commit:
fix: gh workflow

Commit:
fix: gh workflow

Commit:
fix: gh workflow

Commit:
fix: gh workflow

Commit:
fix: gh workflow

Commit:
fix: workingGroup job

Commit:
fix: switch tags

Commit:
fix: gh workflow

Commit:
fix: gh workflow

Commit:
fix: gh workflow

Commit:
fix: gh workflow

Commit:
fix: gh workflow artifacts dir

Commit:
fix: gh workflow error

Commit:
fix: gh artifacts download dir

mnaamani commented on .github/workflows/run-network-tests.yml:
```diff
@@ -185,39 +189,114 @@ jobs:
-    # Re-Enable when we want to test carthage runtime updates
-    if: ${{ false }}
-    name: Runtime Upgrade From Carthage Spec 0
+  build_master_testing:
```
There is some yaml formatting error here that I can't spot that is preventing workflows from running.
Maybe try a yaml format checker ..

Commit:
fix: formatter

Commit:
Merge branch 'ephesus' into ephesus-runtime-upgrade-ci

Commit:
fix: try to fix linter

Commit:
fix: gh workflow file

Commit:
gh workflow

Commit:
fix gh workfow

Commit:
fix: gh workflow docker images saving

Commit:
fix: add docker caching dir

Commit:
fix: reset docker image names

Commit:
fix: gh workflow cache images

Commit:
fix: wrong input path for docker load images

Commit:
fix: try to use different download path

Commit:
fix: gh workflow naming artifacts

Commit:
fix: gh downloaded artifacts names

Commit:
fix: workflow deps

Commit:
fix: try to add 10 sec wait time

Commit:
fix: change sleep time

Commit:
fix: cleanup

Commit:
fix: checks out master repo

Commit:
fix: remove building joystream/js

Commit:
fix: re-enable this repo

Commit:
fix: error with remoe rpc endpoint

Commit:
fix: change wait time

Commit:
fix: cleanup

Commit:
Merge branch 'ephesus' into ephesus-runtime-upgrade-ci

Commit:
fix: format

mnaamani commented on query-node/run-tests.sh:
```diff
@@ -37,6 +37,6 @@ export IGNORE_HIRED_LEADS=true # this directive is needed to run `full` scenario
 SCENARIO=$1
 # fallback if scenario if not specified
-SCENARIO=${SCENARIO:="content-directory"}
+SCENARIO=${SCENARIO:="full"}
```
why this change? I think the scenario chosen here was perhaps to test only query-node without depending on storage infra

mnaamani replied:
I don't understand why this was changed.

mnaamani commented on tests/network-tests/run-migration-tests.sh:
```diff
@@ -1,87 +1,194 @@
-# Post migration assertions by means of typescript scenarios required
-POST_MIGRATION_ASYNC_ASSERTIONS=${POST_MIGRATION_ASYNC_ASSERTIONS:=$true}
-# source common function used for node setup
-source ./node-utils.sh
```
node-utils.sh is not "sourced" anymore, so we can drop the whole script?

mnaamani replied:
I think this script should be renamed to run-runtime-upgrade-tests.sh its not just about migration really.

mnaamani commented on tests/network-tests/run-migration-tests.sh:
```diff
@@ -1,87 +1,194 @@
+}
+
+# Create a chain spec file
+function create_hex_chain_spec() {
```
"hex" is not very meaningful here. "raw" would make more sense.

mnaamani commented on tests/network-tests/run-migration-tests.sh:
```diff
@@ -1,87 +1,194 @@
+        new \
+        --fund-accounts \
+        --authorities //Alice \
+        --sudo-account ${SUDO_ACCOUNT} \
```
Can we drop sudo related arguments so this workflow would not need to be modified again post ephesus. Perhaps using 'generate' command instead of 'new'. In that case you would need to parse the stdout to get the authority keys and specify the keystore path when running the node.

mnaamani commented on tests/network-tests/run-migration-tests.sh:
```diff
@@ -1,87 +1,194 @@
+    CONTAINER_ID=""
+    export JOYSTREAM_NODE_TAG=${RUNTIME}
+    if [ $TARGET_RUNTIME != $RUNTIME ]; then
+        if ! [[ -f ${DATA_PATH}/chain-spec-raw.json ]]; then
```
I had a bit of a hard time testing when some steps failed. Why not just always clear the data folder and start from scratch.

mnaamani commented on tests/network-tests/run-migration-tests.sh:
```diff
@@ -1,87 +1,194 @@
-    CONTAINER_ID=$(start_node)
+    CONTAINER_ID=""
+    export JOYSTREAM_NODE_TAG=${RUNTIME}
+    if [ $TARGET_RUNTIME != $RUNTIME ]; then
```
I think the script should exit with an error if `$TARGET_RUNTIME == $RUNTIME` it doesn't make sense to continue really, because its a user input mistake.

mnaamani commented on utils/api-scripts/src/fork-off.ts:
```diff
@@ -2,17 +2,19 @@ import { xxhashAsHex } from '@polkadot/util-crypto'
 
 // this might not be of much use
-const provider = new WsProvider(process.env.WS_RPC_ENDPOINT || 'ws://localhost:9944')
+const rpcRemoteEndpoint = cmdArgs[1].toString() || 'ws://localhost:9944'
```
All these globals can be local in main()
And not clear what the comment `"// this might not be of much use"` is for

mnaamani commented on .env:
```diff
@@ -94,3 +94,6 @@ BALANCE_LOCKED=300000000000
 # JOYSTREAM_NODE_TAG=latest
+
+# Whether data migration is needed for the runtime upgrade
+DATA_MIGRATION_NEEDED=$false
```
I didn't find any reference to this variable anywhere.

dobertRowneySr replied:
deleted

mnaamani commented on bin/utils/chain-spec-builder/src/main.rs:
```diff
@@ -61,7 +61,7 @@ impl Into<ChainType> for ChainDeployment {
     }
 }
 
-/// A utility to easily create a testnet chain spec definition with a given set
```
No need to to make any changes to runtime file here, just on principle. This means also the image for ephesus is already built and available from docker hub.

mnaamani commented on tests/network-tests/node-utils.sh:
```diff
@@ -61,8 +61,8 @@ function create_chainspec_file {
 	   --chain-spec-path /data/chain-spec.json \
-	   --initial-balances-path /data/initial-balances.json
-}
+	   # --initial-balances-path /data/initial-balances.json # no balances
```
This whole file can be dropped? It is not imported anywhere.

dobertRowneySr replied:
removed

mnaamani commented on tests/network-tests/run-migration-tests.sh:
```diff
@@ -1,87 +1,194 @@
+mkdir -p ${DATA_PATH}
+
+# The latest docker image tag to use for joystream/node (testing profile)
+RUNTIME=${RUNTIME:=mainnetDev}
```
if we can keep the names of the runtimes general would be better. mainnet doesn't make sense, but maybe "current" and "target".. something that works when this code is on a future branch post ephesus .. and with the comment "The latest docker image tag" it is even more confusing.

mnaamani reviewed (changes requested):
I struggled to test locally, but overall the tests are working successfully.
My comments are mainly improvements and some clarification to make code a bit more maintainable, easier to use to test locally. It should become part of regular tooling used by council/community to test runtime upgrade proposals.

mnaamani commented:
Last comment, it might not be necessary to rebuild the docker image from master in the run-network-tests, it should already be available in docker hub. And it might make sense to skip the test when it is running on the master branch, because the test would probably fail, and wouldn't make sense anyways. (starting runtime and target would be the same)

mnaamani commented on tests/network-tests/src/scenarios/full.ts:
```diff
@@ -43,23 +43,16 @@ scenario('Full', async ({ job, env }) => {
   // Runtime upgrade should always be first job
   // (except councilJob, which is required for voting and should probably depend on the "source" runtime)
   const councilJob = job('electing council', electCouncil)
-  const runtimeUpgradeProposalJob = env.RUNTIME_UPGRADE_TARGET_WASM_PATH
```
Looking at this again, I think this job should be kept as it is designed to test all possible outcomes.
For the runtime upgrade github action specifically, we only need to perform a normal scenario where one proposal is made and passes. So maybe have a flow just for that purpose?

Commit:
fix: address changes

Commit:
fix: address changes pt 2

Commit:
fix: address changes pt 3

Commit:
fix: no runtime upgrade on master

Commit:
fix: add back runtime upgrade to full scenario

Commit:
fix: migration file name on gh workflow

Commit:
fix: pass correctly arguments to node

mnaamani commented on tests/network-tests/run-runtime-upgrade-tests.sh:
```diff
@@ -0,0 +1,195 @@
+RUNTIME=${RUNTIME:=current}
+TARGET_RUNTIME=${TARGET_RUNTIME:=target}
+
+# Initial account balance for sudo account
```
Do we sill need //Alice to have a balance?

mnaamani commented on .github/workflows/run-network-tests.yml:
```diff
@@ -189,39 +189,115 @@ jobs:
-    needs: build_images
+    name: Runtime Upgrade From Mainnet To Ephesus
+    needs: [build_master_testing, build_images]
+    if: github.ref != 'refs/heads/master'
```
previous stage `build_master_testing` can also be similarly optional

mnaamani commented on .github/workflows/run-network-tests.yml:
```diff
@@ -189,39 +189,115 @@ jobs:
-    if: ${{ false }}
-    name: Runtime Upgrade From Carthage Spec 0
-    needs: build_images
+    name: Runtime Upgrade From Mainnet To Ephesus
```
```suggestion
    name: Runtime Upgrade From production runtime
```

mnaamani commented on .github/workflows/run-network-tests.yml:
```diff
@@ -189,39 +189,115 @@ jobs:
+          export RUNTIME_CODE_SHASUM=`scripts/runtime-code-shasum.sh`
+          echo "::set-output name=shasum::${RUNTIME_CODE_SHASUM}"
+
+      - name: Setup cache directory
```
I think we should be able to simply just fetch the image from docker hub because the `joystream-node-docker-dev.yml` workflow pushes all the profile builds when code is committed to a branch.

mnaamani reviewed (changes requested):
The tests are still failing.
Left just a few final change requests to speedup the runtime upgrade flow.

Commit:
fix: script setup

Commit:
fix: scripts

Commit:
Update .github/workflows/run-network-tests.yml

Co-authored-by: Mokhtar Naamani <mokhtar.naamani@gmail.com>

Commit:
fix: workflow file

Commit:
fix: renaming

Commit:
fix: new runtime version

Commit:
fix: moduls prefix and post spec clone forkOffChecks

- module prefix are there to make sure that relevant modules are copied
  into the chainspec (you can find them in the polkadot-ui)
- I have also added a new flow for checking post fork-off state

Commit:
fix: linter

Commit:
fix: runtime upgrade selector

Commit:
fix: failing cleanup

Commit:
fix: add checks

Commit:
fix: storage prefix for pallets

Commit:
fix: cleanup scripts

Commit:
fix: cleanup

Commit:
fix: cleanup scripts 2

Commit:
fix: cleanup

Commit:
fix: re enable workflow

Commit:
fix: cleanup

Commit:
fix: full scenario

Commit:
fix: workflow run

mnaamani commented on tests/network-tests/src/scenarios/full.ts:
```diff
@@ -47,9 +47,7 @@ scenario('Full', async ({ job, env }) => {
-  const coreJob = runtimeUpgradeProposalJob || councilJob
-
-  // All other jobs should be executed after coreJob
+  const coreJob = councilJob || runtimeUpgradeProposalJob
```
This change is not correct.
If a runtime upgrade job is running then the core job is the runtime upgrade job (which itself depends on council job). Or you will have tests potentially running before and after upgrade.

Commit:
fix: coreJob

Commit:
fix: full scenario run test

Commit:
fix: scenario job order

Commit:
fix: hire leads after termination

Commit:
Merge branch 'ephesus' into ephesus-runtime-upgrade-ci

Commit:
fix: serialize inviting members job

Commit:
fix: linter

Commit:
fix: remove terminate leads

Commit:
fix: rearrange jobs

Commit:
fix: serialization

Commit:
fix: restore old execution order

dobertRowneySr commented:
there's a problem with the full scenario itself see: https://github.com/Joystream/joystream/pull/4679
I think we should merge this and then move on in order to fix CI checks as a subsequent PR

mnaamani reviewed (approved):
LGTM
Failing test in full scenario will be addressed separately.
